File: android/app/src/main/kotlin/com/jiuxina/ying/WidgetElementStyle.kt
package com.jiuxina.ying

import android.content.SharedPreferences
import android.view.Gravity
import org.json.JSONObject

internal const val WIDGET_ELEMENT_STYLES_KEY = "widget_element_styles"
internal const val WIDGET_VERTICAL_ALIGN_KEY = "widget_vertical_align"

internal enum class WidgetElementVisible {
    follow,
    show,
    hide,
    ;

    companion object {
        fun fromName(value: String?): WidgetElementVisible =
            entries.firstOrNull { it.name == value } ?: follow
    }
}

internal enum class WidgetElementSize(val multiplier: Float) {
    small(0.8f),
    normal(1.0f),
    large(1.25f),
    xlarge(1.5f),
    ;

    companion object {
        fun fromName(value: String?): WidgetElementSize =
            entries.firstOrNull { it.name == value } ?: normal
    }
}

internal enum class WidgetColorMode {
    primary,
    secondary,
    custom,
    ;

    companion object {
        fun fromName(value: String?): WidgetColorMode =
            entries.firstOrNull { it.name == value } ?: primary
    }
}

internal enum class WidgetAlign {
    start,
    center,
    end,
    ;

    companion object {
        fun fromName(value: String?): WidgetAlign =
            entries.firstOrNull { it.name == value } ?: start
    }
}

internal enum class WidgetVerticalAlign {
    top,
    center,
    bottom,
    ;

    companion object {
        fun fromName(value: String?): WidgetVerticalAlign =
            entries.firstOrNull { it.name == value } ?: center
    }
}

internal data class WidgetElementStyle(
    val visible: WidgetElementVisible = WidgetElementVisible.follow,
    val size: WidgetElementSize = WidgetElementSize.normal,
    val colorMode: WidgetColorMode = WidgetColorMode.primary,
    val color: Int = -1,
    val align: WidgetAlign = WidgetAlign.start,
)

internal fun parseElementStyles(raw: String?): Map<String, WidgetElementStyle> {
    if (raw.isNullOrBlank()) return emptyMap()
    return try {
        val json = JSONObject(raw)
        buildMap {
            json.keys().forEach { key ->
                val value = json.optJSONObject(key) ?: return@forEach
                put(key, parseElementStyle(value))
            }
        }
    } catch (_: Exception) {
        emptyMap()
    }
}

internal fun parseElementStyle(json: JSONObject): WidgetElementStyle {
    return WidgetElementStyle(
        visible = WidgetElementVisible.fromName(json.optString("visible")),
        size = WidgetElementSize.fromName(json.optString("size")),
        colorMode = WidgetColorMode.fromName(json.optString("colorMode")),
        color = json.optInt("color", -1),
        align = WidgetAlign.fromName(json.optString("align")),
    )
}

internal fun parseVerticalAlign(value: String?): WidgetVerticalAlign =
    WidgetVerticalAlign.fromName(value)

internal fun elementStyle(
    data: SharedPreferences,
    id: String,
): WidgetElementStyle? = parseElementStyles(
    data.getString(WIDGET_ELEMENT_STYLES_KEY, ""),
).get(id)

internal fun effectiveVisible(
    style: WidgetElementStyle?,
    followDefault: Boolean,
): Boolean = when (style?.visible) {
    WidgetElementVisible.show -> true
    WidgetElementVisible.hide -> false
    else -> followDefault
}

internal fun customElementColor(style: WidgetElementStyle?): Int? =
    if (style?.colorMode == WidgetColorMode.custom && style.color != -1) {
        style.color
    } else {
        null
    }

internal fun elementColor(
    style: WidgetElementStyle?,
    colors: WidgetTextColors,
    defaultPrimary: Boolean,
): Int {
    val fallback = if (defaultPrimary) colors.primary else colors.secondary
    return when {
        style == null -> fallback
        style.colorMode == WidgetColorMode.custom ->
            customElementColor(style) ?: fallback
        style.colorMode == WidgetColorMode.secondary -> colors.secondary
        else -> colors.primary
    }
}

internal fun elementSizeScale(style: WidgetElementStyle?): Float =
    style?.size?.multiplier ?: 1f

internal fun elementGravity(style: WidgetElementStyle?): Int = when (style?.align) {
    WidgetAlign.center -> Gravity.CENTER_HORIZONTAL
    WidgetAlign.end -> Gravity.END
    else -> Gravity.START
}

internal fun widgetDateRowGravity(style: WidgetElementStyle?): Int =
    elementGravity(style) or Gravity.BOTTOM

internal fun widgetVerticalGravity(align: WidgetVerticalAlign): Int = when (align) {
    WidgetVerticalAlign.top -> Gravity.TOP
    WidgetVerticalAlign.bottom -> Gravity.BOTTOM
    else -> Gravity.CENTER_VERTICAL
}

File: lib/models/widget_element_style.dart
import 'dart:convert';

/// 小部件元素的显隐策略；`follow` 表示沿用现有开关与紧凑尺寸规则。
enum WidgetElementVisible {
  follow,
  show,
  hide;
}

/// 小部件元素字号档位，叠加在全局字号缩放之上。
enum WidgetElementSize {
  small,
  normal,
  large,
  xlarge;

  double get multiplier => switch (this) {
    WidgetElementSize.small => 0.8,
    WidgetElementSize.normal => 1.0,
    WidgetElementSize.large => 1.25,
    WidgetElementSize.xlarge => 1.5,
  };
}

/// 小部件文字颜色来源。
enum WidgetColorMode {
  primary,
  secondary,
  custom;
}

/// 小部件文字水平对齐；内联元素不提供对齐选项。
enum WidgetAlign {
  start,
  center,
  end;
}

/// 单事件小部件整体垂直布局。
enum WidgetVerticalAlign {
  top,
  center,
  bottom;
}

/// 单个小部件元素的样式；元素 ID 见设置页与原生渲染的角色清单。
class WidgetElementStyle {
  const WidgetElementStyle({
    this.visible = WidgetElementVisible.follow,
    this.size = WidgetElementSize.normal,
    this.colorMode = WidgetColorMode.primary,
    this.color = -1,
    this.align = WidgetAlign.start,
  });

  final WidgetElementVisible visible;
  final WidgetElementSize size;
  final WidgetColorMode colorMode;
  final int color;
  final WidgetAlign align;

  WidgetElementStyle copyWith({
    WidgetElementVisible? visible,
    WidgetElementSize? size,
    WidgetColorMode? colorMode,
    int? color,
    WidgetAlign? align,
  }) {
    return WidgetElementStyle(
      visible: visible ?? this.visible,
      size: size ?? this.size,
      colorMode: colorMode ?? this.colorMode,
      color: color ?? this.color,
      align: align ?? this.align,
    );
  }

  Map<String, Object> toJson() => {
    'visible': visible.name,
    'size': size.name,
    'colorMode': colorMode.name,
    'color': color,
    'align': align.name,
  };

  factory WidgetElementStyle.fromJson(Map<String, Object?> json) {
    return WidgetElementStyle(
      visible: _enumByName(
        WidgetElementVisible.values,
        json['visible'],
        WidgetElementVisible.follow,
      ),
      size: _enumByName(
        WidgetElementSize.values,
        json['size'],
        WidgetElementSize.normal,
      ),
      colorMode: _enumByName(
        WidgetColorMode.values,
        json['colorMode'],
        WidgetColorMode.primary,
      ),
      color: (json['color'] as int?) ?? -1,
      align: _enumByName(
        WidgetAlign.values,
        json['align'],
        WidgetAlign.start,
      ),
    );
  }
}

T _enumByName<T extends Enum>(List<T> values, Object? name, T fallback) {
  if (name is String) {
    for (final value in values) {
      if (value.name == name) return value;
    }
  }
  return fallback;
}

/// 把元素样式表编码为协议 JSON；旧数据缺省时由解析方回退默认样式。
String encodeWidgetElementStyles(Map<String, WidgetElementStyle> styles) {
  return jsonEncode({
    for (final entry in styles.entries) entry.key: entry.value.toJson(),
  });
}

/// 解析协议 JSON；空值或损坏数据统一回退空配置。
Map<String, WidgetElementStyle> decodeWidgetElementStyles(Object? raw) {
  if (raw is! String || raw.isEmpty) return const {};
  try {
    final decoded = jsonDecode(raw);
    if (decoded is! Map<String, dynamic>) return const {};
    final result = <String, WidgetElementStyle>{};
    decoded.forEach((key, value) {
      if (value is Map<String, dynamic>) {
        result[key] = WidgetElementStyle.fromJson(value);
      }
    });
    return result;
  } catch (_) {
    return const {};
  }
}
